COPTER: Landing gear mavlink control & remove lock state

Pedals2Paddles opened the pull request:
This does a few things related to landing gear in copter.  What prompted me to do this was adding landing gear functionality to the Solo. I found that there was no mavlink landing gear functionality in ArduPilot.  And that stumbled into the auto deploy vs switch position conundrum.  I think I've come up with a method to safely handle all of this, and make the operation easier and cleaner for the operator.

- **COPTER GCS_Mavlink:** Adds mavlink control of landing gear. This will allow you to control the landing gear via GCS and companion computers communicating via mavlink.  Previously, the only way to control landing gear, was using the CH7/8 options, which is very limiting.
  - Add  MAV_CMD_AIRFRAME_CONFIGURATION (# 2520), which is for landing gear control
  - Param 1 is to select specific gear, which is not used in ArduPilot.
  - Param 2 is deploy/retract. 0 is deploy, 1 is Retract
  - My implementation uses 1 to retract and anything other than 1 to deploy to make it error-safe
.
- **COPTER Landing_Gear:** This allows you to programmatically and safely release the "deployed and locked" state that is set when in an automatic landing mode. This locked state prevents the gear from inadvertently retracting after and auto-deploy if the gear switch happens to still be in the up position.  However, this also means the operator has to cycle the gear switch to down (or send the mavlink or mission down command) in order to regain manual control of the gear.  This is confusing and cumbersome.  So I've modified this to be simpler, safe, and easy to operate.
  - Added neutral to the gear switch handling, and added `gear_sw_up` to know when it's up
  - When exiting a mode that auto-deployed the gear, _it checks if the gear switch is up first_.
  - If the gear switch is up, the gear stays in a locked state and the operator is warned.
  - If the gear switch is neutral or down, the lock is released and the gear will operate normally.
  - This prevents the operator from having to unnecessarily cycle the gear switch or commands.
.
- **AP_LANDING_GEAR:** 
  - Add GCS message for deploy & retract.
  - Add an unlock command. Mainly so we can unlock when exiting an auto landing mode.
  - Add a locked bool so the copter code can see if the gear is locked down or not.
.
- **Changes to existing operation:**
  - If the landing gear automatically deploys in an automatic landing mode, _and the RC gear switch is in the neutral or down position_, you no longer need to cycle the switch to regain manual control of the landing gear after the automatic landing terminates. 
  - If the landing gear automatically deploys in an automatic landing mode, _and the RC gear switch is in the up position_, the gear will remain locked down until you take the switch out of the up position. You can go to neutral or down. Either way, the gear will unlock and resume operating normally.
  - You can now use MAV_CMD_AIRFRAME_CONFIGURATION to retract/deploy the gear without needing the CH7/8 option or any RC switch at all.
  - The existing auto-deploy in RTL or land mode, and auto mission rtl/land remains unchanged and will still happen. 


**And it even worked!**
Mavlink commanded gear up and down verified on Mission Planner. The servo PWM out changed appropriately, and the GCS messages came through to tell the operator what it's doing.
![capture](https://user-images.githubusercontent.com/8234496/29747618-6c78cbd0-8ace-11e7-825a-e973eddf7138.JPG)

Pedals2Paddles commented:
@rmackay9 it would be great if we can get this into Copter 3.5.4

DavidIngraham commented:
Any progress on this? I have a PR ready that adds landing gear support for plane but I'd like to rebase it on this to take advantage of the changes to AP_LandingGear.

magicrub commented on ArduCopter/landing_gear.cpp:
```diff
@@ -32,3 +47,16 @@ void Copter::landinggear_update()
+       (control_mode == RTL && (rtl_state == RTL_LoiterAtHome || rtl_state == RTL_Land || rtl_state == RTL_FinalDescent)) ||
+       (control_mode == AUTO && auto_mode == Auto_Land) ||
+       (control_mode == AUTO && auto_mode == Auto_RTL && (rtl_state == RTL_LoiterAtHome || rtl_state == RTL_Land || rtl_state == RTL_FinalDescent))){
+        return true;
```
This is a mouthful. Can we change it to switch(control_mode)?

Pedals2Paddles replied:
@magicrub As it turns out, there really isn't a better way to write this that I can think of. There's only three control modes. It's the "sub modes" that are wordy, and there isn't a way to make that not wordy. It just is.

magicrub commented:
I love the idea of a generic msg MAV_CMD_AIRFRAME_CONFIGURATION but I don't like how it's being used as a gear-only thing. Can't we make this a generic command with enums?

DavidIngraham commented:
My understanding is that MAV_CMD_AIRFRAME_CONFIGURATION is intended primarily for landing gear. In the common message set docs, the only parameters currently implemented are "Landing gear ID" and "Landing gear position". It seems to be a bit of a misnomer, as it suggests a more generic usage.

lvale commented:
@Pedals2Paddles 

Are you aware that you can do this already ??

http://ardupilot.org/planner/docs/common-mavlink-mission-command-messages-mav_cmd.html#mav-cmd-do-set-servo

Pedals2Paddles commented:
I was vacationing on the beach for the weekend so catching up here now.

The `MAV_CMD_AIRFRAME_CONFIGURATION` is really the only existing mavlink command that has anything to do with gear. I suspect the name was chosen with more in mind than just gear, but that's all it has now.

@magicrub Yes I can reword that to be a little more slick. That's actually the original code, I just moved it a but. But I can clean that up while I'm at it. 

@lvale Yes but that's really just a workaround. That's not actually making use of the ArduPilot landing gear functionality and gets it out of sync. I'd like to avoid that. I want to be able to use the true ArduPilot landing gear control and all it's functionality.

Pedals2Paddles commented:
Rebased since it's been a while.  I believe I've made this about as clean as it's going to get.  I also removed the mission aspect since I'm not confident in doing that right. So this is short and sweet now. Adding mission commands can be another day.

@rmackay9  can you have a look this again? I'd love to get this into 3.5.4 if possible. Lots of Solo users have landing gear now, and mavlink retract/deploy would be a big deal.

Pedals2Paddles commented:
I had to redo a lot of stuff again because other changes got pushed through that conflicted. Can we get this merged? @rmackay9

peterbarker commented on ArduCopter/landing_gear.cpp:
```diff
@@ -31,3 +47,15 @@ void Copter::landinggear_update()
 }
+
+// True when copter is in a mode that should auto-deploy the landing gear
+bool Copter::landinggear_deploy_modes()
```
This should be a call on the current FlightMode object:

```
if (flightmode->landing_gear_should_be_deployed()
```

peterbarker commented on ArduCopter/landing_gear.cpp:
```diff
@@ -12,14 +12,30 @@ void Copter::landinggear_update()
         landinggear.set_position(AP_LandingGear::LandingGear_Deploy_And_Keep_Deployed);
     }
 
+    // If we've exited an auto-deploy mode, check the landing gear switch.
```
transition logic between flight modes should be hooked in the flightmode exit function.

peterbarker commented:
On Sat, 9 Dec 2017, Matt wrote:

> I had to redo a lot of stuff again because other changes got pushed through that conflicted. Can we get this merged? @rmackay9

... and I recall why now this was familiar.. I wrote it last week but 
hadn't PR'd it:

https://github.com/ArduPilot/ardupilot/pull/7387

I like this PR, BTW - particularly telling the user what's going on.  If 
you'd like help getting it into the exit_mode stuff, please let me know.

Pedals2Paddles commented:
Rebased and reworked again due to all the other changes that continued to occur.

Pedals2Paddles commented:
@rmackay9 Updated again, only needed some minor syntax changes.  Can we please get this into 3.6?

Pedals2Paddles commented:
It's been almost a year since I created this, and I've had to update it multiple times over this year for other files changing because nobody is merging it.  Now it needs to be updated again, _for the second time this week_.  I am not going to keep rebasing and modifying it for another year if it's going to be ignored.  Can someone please provide some feedback, or merge it, or reject it, or something??

khancyr commented:
Hello,
I think it is worth having that merge ! 
I will add some test functions a little later and ping Randy.

peterbarker commented on ArduCopter/Copter.h:
```diff
@@ -777,6 +777,8 @@ class Copter : public AP_HAL::HAL::Callbacks {
 
     // landing_gear.cpp
     void landinggear_update();
+    bool gcs_gear_sw_warned = false;
```
No need to initialise here.  It's done magically by the program startup sequence.

peterbarker commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -928,6 +928,26 @@ MAV_RESULT GCS_MAVLINK_Copter::handle_command_long_packet(const mavlink_command_
+                case 0:
+                    copter.landinggear.set_position(AP_LandingGear::LandingGear_Deploy);
+                    return MAV_RESULT_ACCEPTED;
+                    break;
```
No need to break.

peterbarker commented on ArduCopter/GCS_Mavlink.cpp:
```diff
@@ -928,6 +928,26 @@ MAV_RESULT GCS_MAVLINK_Copter::handle_command_long_packet(const mavlink_command_
+            // Param 2: 0 = Deploy, 1 = Retract
+            // For safety, anything other than 1 will deploy
+            switch ((uint8_t)packet.param2) {
+                case 0:
```
Don't need this case.

peterbarker commented on ArduCopter/landing_gear.cpp:
```diff
@@ -18,6 +18,21 @@ void Copter::landinggear_update()
+    if (!flightmode->landing_gear_should_be_deployed() && landinggear.locked()) {
+        if (gear_sw_up) {
+            if (!gcs_gear_sw_warned) {
+                gcs().send_text(MAV_SEVERITY_WARNING, "Check Landing Gear Switch");
```
mavlink is a lossy protocol.  This should probably be emitted every minute or so.  Not a huge thing, 'though.

peterbarker commented on ArduCopter/switches.cpp:
```diff
@@ -370,12 +370,14 @@ void RC_Channel_Copter::do_aux_function(const aux_func_t ch_option, const aux_sw
             switch (ch_flag) {
                 case LOW:
                     copter.landinggear.set_position(AP_LandingGear::LandingGear_Deploy);
+                    copter.gear_sw_up = false;
```
Some time in the future I'll make a `bool rc().get_aux_value(const aux_func_t ch_option, aux_switch_pos_t &ch_flag)` method.  This will save having to keep the switch state around like this.

Pedals2Paddles replied:
I was actually just thinking "I wonder if peter made something to do this with more efficiently". 👍

peterbarker commented on libraries/AP_LandingGear/AP_LandingGear.h:
```diff
@@ -40,6 +41,7 @@ class AP_LandingGear {
 
     /// returns true if the landing gear is deployed
     bool deployed() const { return _deployed; }
+    bool locked() const { return _deploy_lock; }
```
This needs a description!  After reading through this patch quickly once I don't get it :-)

In particular, how does this leave the "Unlocked" state and enter the "Retracted" state after the switch is toggled?

Pedals2Paddles replied:
The gear switch handling in copter switches calls the unlock in the same processes that checks the state and warns the operator.

rmackay9 commented on libraries/AP_LandingGear/AP_LandingGear.cpp:
```diff
@@ -81,6 +85,9 @@ void AP_LandingGear::deploy()
     _deployed = true;
+
+    // GCS message & log
+    gcs().send_text(MAV_SEVERITY_INFO, "Landing Gear Deployed");
```
I'd actually prefer if we don't send out too many send-text messages.  I think we should limit the send texts to when things go wrong.  The problem is that it assumes a link with a decent band width and that's not always the case.

Pedals2Paddles replied:
It is more for the purpose of recording the gear action in the log, which I don't believe is done by any other means.

khancyr replied:
for logging in tlog or dataflash ? For dataflash, is it already logged in landinggear_update(). Even if I think it would be simplier that the library do it directly.

Pedals2Paddles replied:
I'll just delete this portion of it entirely.

Pedals2Paddles commented:
Oh great. More changes requiring rebase and updates. I'm very close to walking away from this at this point. Anything else going to change to make more work for me again by the time I get home tonight??

peterbarker commented:
On Wed, 1 Aug 2018, Matt wrote:

> +    // GCS message & log
> +    gcs().send_text(MAV_SEVERITY_INFO, "Landing Gear Deployed");
> 
> It is more for the purpose of recording the gear action in the log, which I don't believe is done by any other means.

You can log a text message into dataflash with something like 
https://github.com/ardupilot/ardupilot/blob/master/ArduCopter/Log.cpp#L533

Pedals2Paddles commented:
I'll keep that in mind for the future.  In this case, I'm just removing it all together so there is one less thing to further delay and further request changes.

Pedals2Paddles commented:
After discussion with Randy, it was determined the long existing deploy-locked state was no longer even needed.  So I removed it. As such, this silly workaround I had to make to UNLOCK the gear is also no longer needed.

So this PR is now just:
- Mavlink control of landing gear
- Updating AP_LANDING_GEAR and Copter Landing gear remove discontinue the lock state

rmackay9 commented:
Merged, thanks!

Pedals2Paddles commented:
@rmackay9 I did https://github.com/ArduPilot/ardupilot/pull/9330 to backport it into 3.6 for you.
